### ISSUE TITLE
Accessibility evaluation of timeliner

### ISSUE THREAD

**Issue by elynema** [avalonmediasystem/avalon#6314]
# Description

SiteImprove cannot scan timelines (unless it somehow gets ahold of a public one). Since these are user-facing pages, we should do a proactive usability evaluation so that we can fix issues prior to the 4/26/2026 deadline.

# Done Looks Like

- [x] Check with SiteImprove browser plugin
  - [x] Timelines page (lists all timelines)
  - [x] Edit timeline details page
  - [x] Copy / delete timeline popups
  - [x] View timeline page (this will create issues in the timeliner repo, rather than avalon repo)
- [x] Check all pages with at least one other accessibility tool: ANDI bookmarklet or WAVE tool
- [x] attempt to navigate through all pages with keyboard only
- [x] Create issues for any identified problems with WCAG 2.1 AA

**Comment by Dananji** [comment]
#### Timeline index page

| Success criterion | WCAG 2.1 Level | Accessibility checker | Error | Relevant HTML element |
| --- | --- | --- | --- | --- |
| 4.1.2 Name, Role, Value | Level A | SiteImprove & WAVE | Form field missing a label | Tag filter drop-down menu |

Keyboard accessibility: everything is keyboard accessible 🎉 

#### Timeline show page

It seems WAVE does not analyze the content within an iframe on a web page. Therefore WAVE was not used on this page.

| Success criterion | WCAG 2.1 Level | Accessibility checker | Error | Relevant HTML element |
| --- | --- | --- | --- | --- |
| 4.1.2 Name, Role, Value | Level A | SiteImprove & axeDevTools | Buttons must have discernible text | Volume thumb button in the volume slider |
| 4.1.2 Name, Role, Value | Level A | SiteImprove & axeDevTools | Inline frame missing a text alternative | iframe used to render timeline |
| 4.1.2 Name, Role, Value | Level A | SiteImprove & axeDevTools | An element with a role that makes its children presentational should not contain focusable elements | Volume slider and its children |
| 2.1.1 Keyboard | Level A | axeDevTools | Scrollable region must have keyboard access | Video player below the timeline |

Keyboard accessibility:
- cannot reach the time-rail, however a keyboard user is able to change the play-head position through play/pause which is keyboard accessible. Once the play-head is at the desired timestamp, they can use create markers and/or split bubbles functionality using the keyboard to edit the timeline
- can reach and activate (using `Spacebar`) all the buttons within the timeline including edit forms for each marker and bubble in the timeline
- does not trap keyboard focus within a modal (saving, settings modals), once keyboard reaches the last focus-able element within the modal, the next `Tab` key-press moves the focus to the links in the footer of the page

Accessibility best practices
| Message | Accessibility checker | Relevant HTML element |
| --- | --- | --- |
| Text not included in an ARIA landmark | SiteImprove | Text elements for timeline bubbles, duration of the player and `Annotations` text below the timeline |
| Page missing headings | SiteImprove | Show page doesn't have heading element in the DOM which helps screen readers understand structure |
| Content missing after heading | SiteImprove | In the `Annotations` section of the page, the `Annotations` heading is using `<h6>`, while the bubble and marker names below it uses `<h3>` elements. This can confuse screen reader when identifying the DOM structure, as `<h3>` and `<h6>` are not in the expected order  |

#### Timeline edit page
| Success criterion | WCAG 2.1 Level | Accessibility checker | Error | Relevant HTML element |
| --- | --- | --- | --- | --- |
| 1.4.3 Contrast (Minimum) | Level AA | SiteImprove & WAVE | Color contrast (1.84 : 1) doesn't meet min requirement | Help texts for visibility, tags and copy LTI link |
| 1.4.3 Contrast (Minimum) | Level AA | SiteImprove & WAVE | Color contrast (3.76 : 1) doesn't meet min requirement | Copy button for LTI link |
| 4.1.2 Name, Role, Value | Level A | SiteImprove & WAVE| Form field missing a text alternative | `<input>` textbox for  timeline `Name` |
| 4.1.2 Name, Role, Value | Level A | SiteImprove & WAVE | Form field missing a label | `<span>` combobox for `Tags` |
| 4.1.2 Name, Role, Value | Level A | SiteImprove & WAVE | Form field missing a label | `<input>` textbox for `Tags` |

Keyboard accessibility: everything is keyboard accessible 🎉 

Accessibility best practices
| Message | Accessibility checker | Relevant HTML element |
| --- | --- | --- |
| Page does not start with a level 1 heading | SiteImprove & WAVE | Edit details page |
| Overuse of italics | SiteImprove | Help text for visibility, tags fields and copy LTI link |
| Orphaned form label | WAVE | Form labels for `Name`, `Visibility`, `Tags`, and `Source` |

#### Copy/delete Timeline popups
| Success criterion | WCAG 2.1 Level | Accessibility checker | Error | Relevant HTML element |
| --- | --- | --- | --- | --- |
| 4.1.2 Name, Role, Value | Level A | SiteImprove | Button missing a text alternative | Close button of the copy timeline modal |
| 4.1.2 Name, Role, Value | Level A | SiteImprove & WAVE| Form field missing a text alternative | `<input>` textbox for  timeline `Name` |
| ARIA-authoring practices | N/A | SiteImprove & axeDevTool | Close button of the copy timeline modal |
| ARIA-authoring practices | N/A | SiteImprove & axeDevtool | ARIA attribute does not exist | `aria-label` is misspelled for the close button of the copy modal |
| ARIA-authoring practices | N/A | SiteImprove | Text not included in an ARIA landmark | Text and buttons inside the delete popover |

Keyboard accessibility:
- Playlist items container does not have visible focus when using tab navigation. When it is reached using tab navigation there is not visual indicator that the focus is on playlist items container after `Share` button.
- Cannot reach the accordion style panels for `Markers` and `Source Item Details` using keyboard navigation.

Accessibility best practices
| Message | Accessibility checker | Relevant HTML element |
| --- | --- | --- |
| Headings are not structured. One or more heading levels have been skipped. | SiteImprove | Copy timeline modal heading |

**Comment by Dananji** [comment]
I didn't create tickets for issues identified related to accessibility best practices, because this ticket is specifically targeting to achieve WCAG 2.1 Level AA conformance.

Plus, I didn't create a ticket for the one thing that could be an issue related to keyboard accessibility,
- does not trap keyboard focus within a modal (saving, settings modals), once keyboard reaches the last focus-able element within the modal, the next Tab key-press moves the focus to the links in the footer of the page

This could probably be tested more and see what the team think?

**Comment by elynema** [comment]
@Dananji I can't test this right now on avalon-dev until we fix the iframe issues. But I took a look at avalon-staging's timeliner implementation. I'm not too worried about the forms that open up to edit annotations not trapping the keyboard within the form since if you tab out of the form, you can still access the rest of the page functionality just fine.

For the settings modal and the save modal, I do think we should trap the keyboard as the user really can't interact with the rest of the page unless the modal is closed. I also think it's an issue that you can't access the color palette options in the Settings modal via keyboard.

Can you create issues for those modals?

I will create an issue for the "skip to main content" link, which doesn't seem to do anything on the timeliner show page.
